src/sage/matroids/circuits_matroid.pyx

```diff
@@ -0,0 +1,676 @@
+        import sage.matroids.unpickling
+        data = (self._groundset, frozenset(self._C), self.get_custom_name())
+        version = 0
+        return sage.matroids.unpickling.unpickle_circuits_matroid, (version, data)
+
+    # enumeration
+
+    cpdef bases(self) noexcept:
+        r"""
+        Return the bases of the matroid.
+
+        OUTPUT: a SetSystem
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: len(M.bases())
+            6
+        """
+        cdef SetSystem B
+        B = SetSystem(list(self.groundset()))
+        from itertools import combinations
+        for S in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(S):
+                    flag = False
+                    break
+            if flag:
+                B.append(S)
+        return B
+
+    def bases_iterator(self):
+        r"""
+        Return an iterator over the bases of the matroid.
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: it = M.bases_iterator()
+            sage: it.__next__()
+            frozenset({0, 1})
+        """
+        from itertools import combinations
+        for B in combinations(self._groundset, self._matroid_rank):
+            flag = True
+            for C in self.nonspanning_circuits():
+                if C <= set(B):
+                    flag = False
+                    break
+            if flag:
+                yield frozenset(B)
+
+    cpdef circuits(self, k=None) noexcept:
+        """
+        Return the circuits of the matroid.
+
+        INPUT:
+
+        - ``k`` -- an integer (optional); the length of the circuits
+
+        OUTPUT: a SetSystem
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: M.circuits()
+            Iterator over a system of subsets
+        """
+        cdef SetSystem C
+        C = SetSystem(list(self.groundset()))
+        if k:
+            for c in self._k_C[k]:
+                C.append(c)
+        else:
+            for i in self._k_C:
+                for c in self._k_C[i]:
+                    C.append(c)
+        return C
+
+    def circuits_iterator(self, k=None):
+        """
+        Return an iterator over the circuits of the matroid.
+
+        INPUT:
+
+        - ``k`` -- an integer (optional); the length of the circuits
+
+        EXAMPLES::
+
+            sage: from sage.matroids.circuits_matroid import CircuitsMatroid
+            sage: M = CircuitsMatroid(matroids.Uniform(2, 4))
+            sage: sum(1 for C in M.circuits_iterator())
+            4
+        """
+        if k:
```

Comment:
```suggestion
        if k is not None:
```
`k` could be `0`, which would then not fall into this path. Probably good to add a test case for this.

I would actually explicitly list the circuits in the doctest too.

Reply:
Done. Good catch!